db/db_impl.cc

```diff
@@ -715,14 +715,51 @@ Status DBImpl::WriteLevel0Table(volatile MemTable* mem, VersionEdit* edit,
     const Slice min_user_key = meta.smallest.user_key();
     const Slice max_user_key = meta.largest.user_key();
     if (base != NULL) {
-        level = base->PickLevelForMemTableOutput(min_user_key, max_user_key);
+        int level_limit;
+        if (0!=options_.tiered_slow_level && (options_.tiered_slow_level-1)<config::kMaxMemCompactLevel)
+            level_limit=options_.tiered_slow_level-1;
+        else
+            level_limit=config::kMaxMemCompactLevel;
+
+        // remember, mutex is held so safe to push file into a non-compacting level
+        level = base->PickLevelForMemTableOutput(min_user_key, max_user_key, level_limit);
+        if (versions_->IsCompactionSubmitted(level) || !versions_->NeighborCompactionsQuiet(level))
+            level=0;
+
         if (0!=level)
         {
+            Status move_s;
             std::string old_name, new_name;
 
             old_name=TableFileName(options_, meta.number, 0);
             new_name=TableFileName(options_, meta.number, level);
-            s=env_->RenameFile(old_name, new_name);
+            move_s=env_->RenameFile(old_name, new_name);
+
+            if (move_s.ok())
+            {
+                // builder already added file to table_cache with 2 references and
```

Comment:
The new file is not added to the manifest until after this routine returns.  The “edit” parameter gets set at line 767 with the new file number and its level.  No thread can access this file in the table cache because the file number is not known outside this routine.  The “edit” object subsequently gets posted to the manifest.  That is when the file number becomes visible.

> On Aug 8, 2015, at 4:06 PM, Paul A. Place <notifications@github.com <mailto:notifications@github.com>> wrote:
> 
> In db/db_impl.cc <https://github.com/basho/leveldb/pull/163#discussion_r36581285>:
> 
> >              std::string old_name, new_name;
> >  
> >              old_name=TableFileName(options_, meta.number, 0);
> >              new_name=TableFileName(options_, meta.number, level);
> > -            s=env_->RenameFile(old_name, new_name);
> > +            move_s=env_->RenameFile(old_name, new_name);
> > +
> > +            if (move_s.ok())
> > +            {
> > +                // builder already added file to table_cache with 2 references and
> The builder added the file to table_cache with mutex_ unlocked; is there any possibility that another thread could have referenced the file via the table_cache before we get to this point?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/basho/leveldb/pull/163/files#r36581285>.
>